File: 037.rb
# 037
# Sudoku Solver
# 2015-05-14
#####################################################
# @param {Character[][]} board
# @return {Void} Do not return anything, modify board in-place instead.
$candidate = []
def solve_sudoku(board)
  printBoard(board)
  for co in (0...81)
    $candidate[co] = []
    if board[co/9][co%9] == '.'
      for k in (1..9)
        if isValid(board, co, k.to_s)
          $candidate[co] << k
        end
      end
    end
    puts "#{co/9} #{co%9} -> #{$candidate[co].inspect}" if not $candidate[co].empty?
  end

  u = uniqueCandidate()
  while u > -1
    i = u / 9
    j = u % 9
    d = $candidate[u].first
    board[i][j] = d.to_s
    for k in (0...9)
      $candidate[9*i+k].delete(d)
      $candidate[9*k+j].delete(d)
      $candidate[9*(i/3*3+k/3)+j/3*3+k%3].delete(d)
    end
    u = uniqueCandidate()
  end
  for co in (0...81)
    puts "#{co/9} #{co%9} -> #{$candidate[co].inspect}" if not $candidate[co].empty?
  end

  solve(board, 0)
  printBoard(board)
  return
end
def solve(board, co)
  return true if co == 81
  i = co / 9
  j = co % 9
  return solve(board, co+1) if board[i][j] != '.'
  for k in (1..9)
    if isValid(board, co, k.to_s)
      board[i][j] = k.to_s
      if solve(board, co+1)
        return true
      else
        board[i][j] = '.'
      end
    end
  end
  false
end
def isValid(board, co, value)
  i = co / 9
  j = co % 9
  for k in (0...9)
    return false if (board[i][k] == value or board[k][j] == value or board[i/3*3+k/3][j/3*3+k%3] == value)
  end
  true
end
def uniqueCandidate()
  for co in (0...81)
    return co if $candidate[co].size == 1
  end
  -1
end


def printBoard(board)
  for i in (0...board.size())
    for j in (0...board.size())
      print board[i][j]
      print '  ' if j % 3 == 2
    end
    print "\n"
    print "\n" if i % 3 == 2
  end
end

#####################################################

board = [
"..9748...",
"7........",
".2.1.9...",
"..7...24.",
".64.1.59.",
".98...3..",
"...8.3.2.",
"........6",
"...2759.."
]

solve_sudoku(board)

File: python/037.py
#   037
#   Sudoku Solver
#   2015-05-14
#####################################################
import math
class Solution:
    candidate = []
    # @param {character[][]} board
    # @return {void} Do not return anything, modify board in-place instead.
    def solveSudoku(self, board):
        printBoard(board)

        for co in range(81):
            i = co // 9
            j = co % 9
            self.candidate.append(0)
            if board[i][j] == '.':
                for k in range(1, 10):
                    if self.isValid(board, co, chr(ord('0')+k)):
                        self.candidate[co] |= 1<<k
            if self.candidate[co]:
                print(i, j, self.candidate[co])

        u = self.uniqueCandidate(board)
        while u > -1:
            i = u // 9
            j = u % 9
            d = int(math.log(self.candidate[u], 2))
            board[i][j] = chr(ord('0')+d)
            print(i, j, " <- ", d)
            for k in range(9):
                self.candidate[9*i+k] &= ~(1<<d)
                self.candidate[9*k+j] &= ~(1<<d)
                self.candidate[9*(i//3*3+k//3)+j//3*3+k%3] &= ~(1<<d)
            u = self.uniqueCandidate(board)
        print()
        for co in range(81):
            if self.candidate[co]:
                print(co//9, co%9, self.candidate[co])

        self.solve(board, 0)
        printBoard(board)
        return
    def solve(self, board, co):
        if co == 81:
            return True
        i = co // 9
        j = co % 9
        if board[i][j] != '.':
            return self.solve(board, co+1)
        for k in range(1, 10):
            if self.isValid(board, co, chr(ord('0')+k)):
                board[i][j] = chr(ord('0')+k)
                if self.solve(board, co+1):
                    return True
                board[i][j] = '.'
        return False
    def isValid(self, board, co, c):
        i = co // 9
        j = co % 9
        for k in range(9):
            if board[i][k] == c or board[k][j] == c or board[i//3*3+k//3][j//3*3+k%3] == c:
                return False
        return True
    def uniqueCandidate(self, board):
        for co in range(81):
            if self.candidate[co] in [2, 4, 8, 16, 32, 64, 128, 256, 512]:
                return co
        return -1



def printBoard(board):
    for i in range(9):
        for j in range(9):
            print(board[i][j], end='')
            if j % 3 == 2:
                print(' ', end='')
        print()
        if i % 3 == 2:
            print()

#####################################################

board = [
"..9748...",
"7........",
".2.1.9...",
"..7...24.",
".64.1.59.",
".98...3..",
"...8.3.2.",
"........6",
"...2759.."
]

board = [
"53..7....",
"6..195...",
".98....6.",
"8...6...3",
"4..8.3..1",
"7...2...6",
".6....28.",
"...419..5",
"....8..79"
]

for i in range(9):
    board[i] = list(board[i])

sol = Solution()
sol.solveSudoku(board)

